src/python/pants/backend/python/subsystems/python_tool_base.py

```diff
@@ -143,6 +143,33 @@ def pex_requirements(self) -> PexRequirements:
             is_lockfile=True,
         )
 
+    def pex_requirements_with_digest(self, expected_lockfile_digest: str) -> PexRequirements:
+        """The requirements to be used when installing the tool.
+
+        If the tool supports lockfiles, the returned type will install from the lockfile rather than
+        `all_requirements`.
+        """
```

Comment:
Another factoring might be: update `pex_requirements` to be a method with `expected_lockfile_digest: str | None = None`. If `self.register_lockfile and self.lockfile != "<none>", raise an AssertionError if the kwarg is not provided.

That avoids having two separate functions, which might make things more clear. I'm concerned about people not using _with_digest when they need to be.

> `self.lockfile != "<none>"`:

This is how you disable lockfiles. If that happens, we should not waste time computing what the hash would be.

Reply:
Yup, that sounds reasonable!

Reply:
@Eric-Arellano It looks like I'll need to adjust [this section](https://github.com/pantsbuild/pants/blob/8a0849c508ab88f1cda1a6c54f7502783f5e1988/src/python/pants/backend/python/subsystems/python_tool_base.py#L129-L144) as well -- is it just a matter of using `self.interpreter_constraints` and `self.all_requirements`?

Reply:
(deleted previous comment; actually read the code)